pkg/network/status/status_test.go

```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		t.Fatalf("IsReady failed: %v", err)
+	}
+	if ok {
+		t.Fatalf("IsReady returned %v, want: %v", ok, false)
+	}
+
+	// Wait for the first request (success) to be executed
+	<-probeRequests
+
+	// Wait for the probing to eventually succeed
+	<-ready
+
+	select {
+	// Validate that no requests went through the probe handler
+	case <-dummyRequests:
+		t.Fatal("an unexpected request went through the probe handler")
+	default:
+		break
+	}
+}
+
+func TestCancellation(t *testing.T) {
+	ia := iaTemplate.DeepCopy()
+	// Handler keeping track of received requests and mimicking an Ingress not ready
+	requests := make(chan *http.Request, 100)
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		requests <- r
+		w.WriteHeader(404)
+	})
+
+	ts := httptest.NewServer(handler)
+	defer ts.Close()
+	url, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("failed to parse URL %q: %v", ts.URL, err)
+	}
+	hostname := url.Hostname()
+	if err != nil {
+		t.Fatalf("failed to parse port %q: %v", url.Port(), err)
+	}
+	port, err := strconv.Atoi(url.Port())
+	if err != nil {
+		t.Fatalf("failed to parse port %q: %v", url.Port(), err)
+	}
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "gateway",
+		},
+		Status: v1.PodStatus{
+			PodIP: strings.Split(url.Host, ":")[0],
+		},
+	}
+
+	ready := make(chan *v1alpha1.Ingress)
+	prober := NewProber(
+		zaptest.NewLogger(t).Sugar(),
+		&fakeProbeTargetLister{
+			map[string]map[string]sets.String{
+				hostname: map[string]sets.String{
+					strconv.Itoa(port): sets.NewString("http://foo.bar.com:80"),
+				},
+			},
+		},
+		func(ia *v1alpha1.Ingress) {
+			ready <- ia
+		})
+
+	done := make(chan struct{})
+	defer close(done)
+	prober.Start(done)
+
+	ok, err := prober.IsReady(context.Background(), ia)
+	if err != nil {
+		t.Fatalf("IsReady failed: %v", err)
+	}
+	if ok {
+		t.Fatalf("IsReady returned %v, want: %v", ok, false)
+	}
+
+	// Wait for the first probe request
+	<-requests
+
+	// Create a new version of the Ingress
+	ia = ia.DeepCopy()
+	ia.Spec.Rules[0].Hosts[0] = "blabla.net"
+
+	// Check that probing is unsuccessful
+	select {
+	case <-ready:
+		t.Fatal("Probing succeeded while it should not have succeeded")
+	default:
+	}
+
+	ok, err = prober.IsReady(context.Background(), ia)
+	if err != nil {
+		t.Fatalf("IsReady failed: %v", err)
+	}
+	if ok {
+		t.Fatalf("IsReady returned %v, want: %v", ok, false)
+	}
+
+	// Drain requests for the old version
+	for req := range requests {
+		log.Printf("req.Host: %s", req.Host)
+		if strings.HasPrefix(req.Host, "blabla.net") {
+			break
+		}
+	}
+
+	// Cancel Pod probing
+	prober.CancelPodProbing(pod)
+
+	// Check that the requests were for the new version
+	close(requests)
+	for req := range requests {
+		if !strings.HasPrefix(req.Host, "blabla.net") {
+			t.Fatalf("Unexpected Host: want: %s, got: %s", "blabla.net", req.Host)
+		}
+	}
+}
+
+type fakeProbeTargetLister struct {
```

Comment:
why do you need the struct?
Just `type fakePTL map[string]map[string]sets.String`.

Reply:
Done